lib/change_stream.js

```diff
@@ -534,8 +494,112 @@ function processNewChange(args) {
   changeStream.options.startAtOperationTime = undefined;
 
   // Return the change
-  if (eventEmitter) return changeStream.emit('change', change);
-  return callback(error, change);
+  if (!callback) return changeStream.emit('change', change);
+  return callback(undefined, change);
+}
+
+function processError(changeStream, error, callback) {
+  const topology = changeStream.topology;
+  const cursor = changeStream.cursor;
+
+  // If the change stream has been closed explictly, do not process error.
+  if (changeStream.closed) {
+    if (callback) callback(new MongoError('ChangeStream is closed'));
+    return;
+  }
+
+  // if the resume succeeds, continue with the new cursor
+  function resumeWithCursor(newCursor) {
+    changeStream.cursor = newCursor;
+    processResumeQueue(changeStream);
+  }
+
+  // otherwise, raise an error and close the change stream
+  function unresumableError(err) {
+    if (!callback) {
+      changeStream.emit('error', err);
+      changeStream.emit('close');
+    }
+    processResumeQueue(changeStream, err);
+    changeStream.closed = true;
+  }
+
+  if (cursor && isResumableError(error, maxWireVersion(cursor.server))) {
+    changeStream.cursor = undefined;
+
+    // stop listening to all events from old cursor
+    ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
+
+    // close internal cursor, ignore errors
+    cursor.close();
+
+    // the presence of a callback here indicates the error happened during a .next
+    // so it should be retried on resume
+    if (callback) changeStream[kResumeQueue].push(() => changeStream.next(callback));
+
+    waitForTopologyConnected(topology, { readPreference: cursor.options.readPreference }, err => {
+      // if the topology can't reconnect, close the stream
+      if (err) return unresumableError(err);
+
+      // create a new cursor, preserving the old cursor's options
+      const newCursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
+
+      // attempt to continue in emitter mode
+      if (!callback) return resumeWithCursor(newCursor);
+
+      // attempt to continue in iterator mode
+      newCursor.hasNext(err => {
+        // if there's an error immediately after resuming, close the stream
+        if (err) return unresumableError(err);
+        resumeWithCursor(newCursor);
+      });
+    });
+    return;
+  }
+
+  if (!callback) return changeStream.emit('error', error);
+  return callback(error);
+}
+
+/**
+ * Safely provides a cursor across resume attempts
+ *
+ * @param {ChangeStream} changeStream the parent ChangeStream
+ * @param {function} callback gets the cursor or error
+ * @param {ChangeStreamCursor} [oldCursor] when resuming from an error, carry over options from previous cursor
+ */
+function getCursor(changeStream, callback) {
+  if (changeStream.isClosed()) {
```

Comment:
just want to call this out here, but you are mixing checks for being closed between using the `isClosed` method and directly checking the `closed` boolean (in `processError`, `processNewChange`). Just want to make sure this intentional, and see if we can maybe consolidate to one approach. It occurs to me it might be because `isClosed` considers the existence of the `cursor`, so there might be some very subtle reasons to use one over the other

Reply:
Yeah that's precisely the reason. I could consolidate them by doing something like this:
```
  /**
   * Is the change stream closed
   * @method ChangeStream.prototype.isClosed
   * @param  {boolean} checkCursor also check if the underlying cursor is closed
   * @return {boolean}
   */
  isClosed(checkCursor) {
    return this.closed || (checkCursor && this.cursor && this.cursor.isClosed());
  }
```
then using `isClosed(true)` for the current `isClosed()` checks, and `isClosed()` for the current checks against the `.closed` property.

Reply:
The other direction to go would be to just inline the checks for the closed cursor in the two places they are needed, and only check the `closed` property in `isClosed`.